util: update rbd and cephfs driver volume capability to support SINGLE_NODE_{SINGLE/MULTI}_WRITER and support RWOP PVCs

humblec opened the pull request:
SINGLE_NODE_WRITER capability ambiguity has been fixed in csi spec v1.5
which allows the SP drivers to declare more granular WRITE capability in form
of SINGLE_NODE_SINGLE_WRITER or SINGLE_NODE_MULTI_WRITER.
These are not really new capabilities rather capabilities introduced to
get the desired functionality from CO side based on the capabilities SP
driver support for various CSI operations, this new capabilities also help to address
new access mode RWOP ( readwriteoncepod) based on further requirement.

This patch adds or make our CSI driver to also declare these capabilities. 
This functionality required "ReadWriteOncePod" feature gate to be enabled in the kube cluster.


Updates: https://github.com/ceph/ceph-csi/pull/2357

Additional Ref # https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

What is covered in this PR?
- [x] Advertising the new capabilities from the controller, node servers based on csi spec 1.5 
- [x] RBD controller server updates for new access mode checks
- [x] Introduction of new helper function to validate capabiliites 
- [x] Preserving  backward compatibility of RWO File mode PVCs
- [x] E2E tests for RWOP access mode RBD PVCs for both Block and File mode
- [x] Example Yamls with new access mode
- [x] CephFS e2e tests
- [x] Documentation update for RWOP accessmode support

humblec commented:
/retest all

Madhu-1 commented on examples/rbd/pvc.yaml:
```diff
@@ -5,7 +5,7 @@ metadata:
 spec:
   accessModes:
-    - ReadWriteOnce
+    - ReadWriteOncePod
```
please add a new example instead of changing the existing one.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -101,7 +101,8 @@ func (cs *ControllerServer) parseVolCreateRequest(
 	for _, capability := range req.VolumeCapabilities {
 		// RO modes need to be handled independently (ie right now even if access mode is RO, they'll be RW upon attach)
-		if capability.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER {
+		if (capability.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER) ||
```
Add a new test case for new accessMode

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -120,7 +121,7 @@ func (cs *ControllerServer) parseVolCreateRequest(
 	}
 
-	// if it's NOT SINGLE_NODE_WRITER and it's BLOCK we'll set the parameter to ignore the in-use checks
+	// if it's NOT SINGLE_NODE_SINGLE_WRITER and it's BLOCK we'll set the parameter to ignore the in-use checks
```
why `SINGLE_NODE_WRITER` is removed from here?

nixpanic replied:
I wonder about this too. Is `SINGLE_NODE_WRITER` deprecated and replaced by `SINGLE_NODE_SINGLE_WRITER` maybe? If so, the commit message should explain that clearly.

Yuggupta27 replied:
If it is depreciated, then we need to get rid of  https://github.com/ceph/ceph-csi/pull/2359/commits/cfe8f83c72724fe88f658943995dd97ea8098a11#diff-4bc3ac45c059d328076214abbf3b72db3cd475699833d9825fcaa6471a6ccd85R147 too.

Madhu-1 reviewed (dismissed):
@humblec Please update the PR description/commit message without that its difficult to understand why this change is required and the also same changes are not required for cephfs?

nixpanic reviewed (dismissed):
Definitely missing a comment in the commit that describes this change and the need for it.

It is unclear to me why this `Updates: #2358`, maybe you mean `Depends-on: #2358` instead?

Obviously existing examples and test-cases should not be adjusted for new features, instead new examples and tests need to be added.

humblec commented:
> Definitely missing a comment in the commit that describes this change and the need for it.
> 
> It is unclear to me why this `Updates: #2358`, maybe you mean `Depends-on: #2358` instead?
> 
> Obviously existing examples and test-cases should not be adjusted for new features, instead new examples and tests need to be added.

@Madhu-1 @nixpanic sure,  I made a mistake by not placing `WIP` here, let me revisit this PR.

nixpanic commented:
> @Madhu-1 @nixpanic sure,  I made a mistake by not placing `WIP` here, let me revisit this PR. 

Just create draft PRs for things that are still a work-in-progress. No
need to set/remove labels that way, any contributor can use this
workflow.

humblec committed:
cephfs: add SINGLE_NODE_{SINGLE/MULTI}_WRITER capability

SINGLE_NODE_WRITER capability ambiguity has been fixed in csi spec v1.5
which allows the SP drivers to declare more granular WRITE capability.
These are not really new capabilities rather capabilities introduced to
get the desired functionality from CO side based on the capabilities SP
driver support for various CSI operations.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest all

humblec commented:
/retest  ci/centos/mini-e2e/k8s-1.21

humblec commented:
/retest all

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -97,20 +97,23 @@ func (cs *ControllerServer) parseVolCreateRequest(
+		// below capability check indicates that we support both {SINGLE_NODE or MULTI_NODE} WRITERs and the `isMultiWriter`
+		// flag has been set accordingly.
+		if (capability.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER) ||
+			(capability.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER) {
```
question:
if I read this currently we are not allowing rbd filesystem mode pvc on a single node with multiple pods `VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER`  is that correct?

RBD filesystem RWO should be supported?

humblec replied:
Yes, your reading is correct, we are not allowing fs mode PVC on more than one writer regardless of rwo or rwx.

Madhu-1 reviewed (dismissed):
Please add the example and test cases for both RBD (file and block Mode) and Cephfs to make sure the functionality is working fine

* Create an RWOP PVC
* Bind it a application pod (this should successful)
* Try to Bind this to another application pod (this should fail)

Madhu-1 commented:
@mergifyio rebase

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -98,20 +98,23 @@ func (cs *ControllerServer) parseVolCreateRequest(
 		// RO modes need to be handled independently (ie right now even if access mode is RO, they'll be RW upon attach)
-		if capability.GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER {
-			isMultiNode = true
+		// below capability check indicates that we support both {SINGLE_NODE or MULTI_NODE} WRITERs and the `isMultiWriter`
```
Might be useful to add a `isMultiWriter([]VolumeCapability)` function so that comments and long if-statements do not clutter this function so much. It will also make it easy to add a unit-test for the checks on the new capability without regressing on the previous behaviour.

humblec replied:
Sure, that can be added.

humblec replied:
Done.

humblec commented:
as we have 1.22 CI tests enforced now in this repo, let me refresh this PR with RWOP.. thanks :+1:

humblec committed:
rbd: add SINGLE_NODE{SINGLE_MULTI}_WRITER capabilities

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
rbd: introduce a helper function to detect multi writer,block & rwofile

SINGLE_NODE_WRITER capability ambiguity has been fixed in csi spec v1.5
which allows the SP drivers to declare more granular WRITE capability in form
of SINGLE_NODE_SINGLE_WRITER or SINGLE_NODE_MULTI_WRITER.

These are not really new capabilities rather capabilities introduced to
get the desired functionality from CO side based on the capabilities SP
driver support for various CSI operations, this new capabilities also help
to address new access mode RWOP (readwriteoncepod).

This commit adds a helper function which identity the request is of
multiwriter mode and also validates whether it is filesystem mode or
block mode. Based on the inspection it fails to allow multi write
requests for filesystem mode and only allow multi write request against
block mode.

This commit also adds unit tests for isMultiWriterBlock function which
validates various accesstypes and accessmodes.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
hmmmm.. looks like minikube is not picking up the featuregate properly for older Kube releases <1.22.

humblec commented:
@nixpanic @Madhu-1 all the comments are addressed and also added e2e tests.  Please review. 

Additional note: Currently e2e tests are passing for kube 1.22 version, however minikube fails to come up while it deploy kube 1.21 or older versions, this looks like because of the new feature gate introduced only in Kube 1.22.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,77 @@ var _ = Describe("RBD", func() {
+				if err != nil {
+					e2elog.Failf("failed to delete application with error %v", err)
+				}
+				app.Name = aname
```
This can be added to the above test no need to have an extra test for this one.

humblec replied:
ok.. can be clubbed.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,77 @@ var _ = Describe("RBD", func() {
+				err = deletePVCAndApp("", f, pvc, app)
+				if err != nil {
+					e2elog.Failf("failed to delete PVC and application with error %v", err)
+				}
```
validate the image in pool and trash

humblec replied:
sure, done.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,77 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+			})
+			By("create a RWOP PVC and bind it to more than an app", func() {
```
move this to util and reuse same for cephfs as no e2e is added for cephfs

humblec replied:
@Madhu-1 rather I will add e2e tests for cephfs.  as it ensures we are not breaking on CephFS side for various access mode.

Madhu-1 replied:
not sure how this helps. the code block remains the same for validating the RWOP access mode. instead duplicating the code we can move the generic validation to a common helper function and reuse that for rbd and cephfs to avoid duplication

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -94,25 +94,39 @@ func (cs *ControllerServer) validateVolumeReq(ctx context.Context, req *csi.Crea
+		if cap.AccessMode != nil {
+			// we check the capability access mode is multi writer regardless of it is single or multi node.
+			if (cap.AccessMode.Mode == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER) ||
+				(cap.AccessMode.Mode == csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER) {
```
Does this mean we are blocking filesystem PVC creation for RWO?

humblec replied:
shouldnt be . Also all e2e tests ( except raw block) proves that?

Madhu-1 replied:
am not checking on the E2E am mainly going by code reading if the accesMode is `singleNode multi writer` you are considering it as a multiwrite=true and Block will be false and considering below check

```
if isMultiWriter && !isBlock {
		return nil, status.Error(
			codes.InvalidArgument,
			"multi node access modes are only supported on rbd `block` type volumes")
```
If its RWO (which is singleNodeMultiWriter) cephcsi will return error message

humblec replied:
Indeed thats what I meant, that said, if there are multiple writers to the volume of FS mode, I was blocking that operation considering possibilities of corruption. Now there are a couple of thoughts on this, ie the corruption scenario is limited in this situation of SINGLE node multi writer, so may be with a disclaimer we can allow that and kind of preserve the backward compatibility. But, I have to do some more data integrity testing before that.

Madhu-1 replied:
you mentioned you are not blocking RWO filesystem PVC and all E2E tests are passing here https://github.com/ceph/ceph-csi/pull/2359#discussion_r737120013. it was not a problem before till now in cephcsi we allowed this operation. I would suggest opening a new issue with details about this one.

humblec replied:
yeah, RWO mode was a relaxed one so far. this is more strict, but the point here is that, the backward compatibility. some discussions happened on the same and we can expect data corruption if different apps are operating on the same file/dir in rwo scenario, again, almost we are concluding that, we can allow RWO mode but can have a disclaimer in the doc. so I will refresh this PR with allowing that and documentation can cover it. Thats the plan.

Madhu-1 replied:
>some discussions happened on the same and we can expect data corruption if different apps are operating on the same file/dir in rwo scenario

please provide a link to this one so that helps in understanding more about it.

> but the point here is that, the backward compatibility

what is the backward compatibility here? please elaborate more about it in the commit message for reference.

humblec replied:
> some discussions happened on the same and we can expect data corruption if different apps are operating on the same file/dir in rwo scenario
>>please provide a link to this one so that helps in understanding more about it.

This was discussed in CSI sync up call  (looks like you were absent ) where along with other folks who were in,  we had a long discussion where @pkalever and @nixpanic extensively talk about these scenarios.

Madhu-1 replied:
>we had a long discussion where @pkalever and @nixpanic extensively talk about these scenarios

@humblec Thanks for the details. If it was an extensive long discussion i assume all the technical aspects are discussed and we are completely settled on the things. If yes, I would expect this summary to be captured in a Github issue or PR   for the people who are not part of the meeting and for upstream reference.

humblec replied:
Summary is the earlier comment https://github.com/ceph/ceph-csi/pull/2359#discussion_r739009748

Madhu-1 replied:
The comment does not provide complete insights (which is not useful at least for me) am more looking for discussion details on data corruption and when it can happen? current we are supporting RWO for a long time I haven't seen any data corruption.

pkalever replied:
I guess the discussion on the other day of more on the RWX rather than on RWO, more specifically about the filesystem volume mode. At the end of the discussions, we had two opinions about allowing {RWX + SINGLE NODE + FS} case

1. We must continue to restrict, MULTINODE MULTIWRITERS for FS Mode. But allow the same for SINGLENODE MULTIWRITERS with some docs and options to enable IMHO (default: disable)
2. @nixpanic concurs with the above but he suggested we don't need any documentation to allow RWX on a single node as this case looks identical to RWO (Niels can still correct me here)

I guess we take it for granted that every admin will be aware of corruption involved in accessing the same file from multiple apps/pods when application-level fencing doesn't exist.

humblec replied:
@pkalever pretty much concluded, but a minor clarification on `RWX rather than on RWO` -> this was actually `rwo + multi writer` scenario.. rest of the above should carry.

humblec replied:
``` current we are supporting RWO for a long time I haven't seen any data corruption.``` -> The data corruption can occur and RWO was **not** enough to properly respect the intention meant by users with this accessmode, so this change or feature got introduced in kubernetes in first place .

Madhu-1 commented on internal/rbd/controllerserver_test.go:
```diff
@@ -67,3 +67,114 @@ func TestIsThickProvisionRequest(t *testing.T) {
+			}
+			if got1 != newtt.block {
+				t.Errorf("isMultiWriterBlock() got1 = %v, want %v", got1, newtt.block)
+			}
```
got and got1 can be renamed to more readable variable names

humblec replied:
done.

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -160,7 +160,7 @@ CSI_RESIZER_VERSION=${CSI_RESIZER_VERSION:-"v1.2.0"}
 
 #feature-gates for kube
-K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"ExpandCSIVolumes=true"}
+K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"ExpandCSIVolumes=true,ReadWriteOncePod=true"}
```
this need to be adjusted to set only if kubernetes version supports it

humblec replied:
done.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -54,6 +54,7 @@ var (
 	pvcClonePath           = rbdExamplePath + "pvc-restore.yaml"
 	pvcSmartClonePath      = rbdExamplePath + "pvc-clone.yaml"
 	pvcBlockSmartClonePath = rbdExamplePath + "pvc-block-clone.yaml"
+	pvcRWOPPath            = rbdExamplePath + "pvc-rwop.yaml"
```
do we have this file already created?

humblec replied:
yes. thanks.

humblec commented:
/retest all

humblec commented:
hmmm.. 1.22 also fails with minikube deployment changes..

humblec commented:
@Madhu-1 @nixpanic ptal.. thanks

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,52 @@ var _ = Describe("RBD", func() {
 				// validate created backend rbd images
 				validateRBDImageCount(f, 0, defaultRBDPool)
 			})
+			By("create a RWOP PVC and bind it to more than an app", func() {
```
```suggestion
create a RWOP PVC and bind it to more than one app
```

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,52 @@ var _ = Describe("RBD", func() {
+				}
+				pvc, err := loadPVC(pvcRWOPPath)
+				if err != nil {
+					e2elog.Failf("failed to load PVC with error %v", err)
```
please remove ` with error %v` from the error message. mostly in all places which are getting added in this PR

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -420,6 +421,52 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("failed to delete PVC and application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0, defaultRBDPool)
```
please add check to validate image in trash

pkalever replied:
* Can we perform some IO here and prove the behaviors from different apps? App having Write access should succeed and rest should expect a ReadOnlyErr.
* Can we add another test case for Block volumeMode?

humblec replied:
@pkalever i/o may not help as the scheduling itself is blocked for the second app, but I can add the block mode test.

Madhu-1 commented on scripts/minikube.sh:
```diff
@@ -188,6 +197,7 @@ if [[ "${KUBE_VERSION}" == "latest" ]]; then
     KUBE_VERSION=$(curl -L https://storage.googleapis.com/kubernetes-release/release/stable.txt 2> /dev/null)
 fi
 
+
```
remove this extra line

Madhu-1 commented on internal/cephfs/driver.go:
```diff
@@ -107,6 +107,8 @@ func (fs *Driver) Run(conf *util.Config) {
 		fs.cd.AddVolumeCapabilityAccessModes([]csi.VolumeCapability_AccessMode_Mode{
 			csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
+			csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+			csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER,
```
```text
// Indicates the SP supports the SINGLE_NODE_SINGLE_WRITER and/or
	// SINGLE_NODE_MULTI_WRITER access modes.
	// These access modes are intended to replace the
	// SINGLE_NODE_WRITER access mode to clarify the number of writers
	// for a volume on a single node. Plugins MUST accept and allow
	// use of the SINGLE_NODE_WRITER access mode when either
	// SINGLE_NODE_SINGLE_WRITER and/or SINGLE_NODE_MULTI_WRITER are
	// supported, in order to permit older COs to continue working.
	ControllerServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER ControllerServiceCapability_RPC_Type = 13
```
any specific reason not to expose https://github.com/container-storage-interface/spec/blob/5b0d4540158a260cb3347ef1c87ede8600afb9bf/lib/go/csi/csi.pb.go#L238-L246 this controller capabilities?

Madhu-1 replied:
without added new controller capability external-provisioner will not consider sending new accessMode to csi driver https://github.com/kubernetes-csi/external-provisioner/blob/3a77c811ff2166e4642252c48ee79dfb3b584a8e/pkg/controller/controller.go#L484-L488

Madhu-1 reviewed (dismissed):
a couple of things are still pending

* E2E for cephfs
* Update of the readme for new access mode

Madhu-1 commented:
This is from my local test of these changes + exposing the new controller caps. not sure why it is not considered

```
---yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rbd-pvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: rook-ceph-block
```
```bash
kubectl descectl describe pvc
Name:          rbd-pvc
Namespace:     default
StorageClass:  rook-ceph-block
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: rook-ceph.rbd.csi.ceph.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                Age                     From                                                                                                        Message
  ----     ------                ----                    ----                                                                                                        -------
  Normal   ExternalProvisioning  2m49s (x26 over 8m36s)  persistentvolume-controller                                                                                 waiting for a volume to be created, either by external provisioner "rook-ceph.rbd.csi.ceph.com" or manually created by system administrator
  Normal   Provisioning          4s (x11 over 8m36s)     rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-5845764558-49sdn_646d5259-1d31-47bc-94ab-31c1033311eb  External provisioner is provisioning volume for claim "default/rbd-pvc"
  Warning  ProvisioningFailed    4s (x11 over 8m36s)     rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-5845764558-49sdn_646d5259-1d31-47bc-94ab-31c1033311eb  failed to provision volume with StorageClass "rook-ceph-block": rpc error: code = InvalidArgument desc = multi node access modes are only supported on rbd `block` type volumes
```

humblec commented:
>....
  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-5845764558-49sdn_646d5259-1d31-47bc-94ab-31c1033311eb  failed to provision volume with StorageClass "rook-ceph-block": rpc error: code = InvalidArgument desc = multi node access modes are only supported on rbd `block` type volumes
>

This is what the expected behavior with the new changes and logic in place. But  based on the discussion pointed out in other thread, to preserve backward compatibility current patch version has been modified to allow RWO file mode.

humblec commented:
Appreciate review on this as 90% of the requirement is covered already as mentioned in the PR description. 

Cc @ceph/ceph-csi-contributors

humblec commented:
Test failure looks unrelated https://github.com/ceph/ceph-csi/issues/2677

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@ceph/ceph-csi-contributors can you please review this PR?

humblec commented:
@Mergifyio rebase

humblec commented:
@ceph/ceph-csi-contributors can you review this ?

Rakshith-R commented on scripts/minikube.sh:
```diff
@@ -203,6 +212,14 @@ up)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 22 ];
+    then
+    # if kubernetes version is greater than 1.22 enable RWOP feature gate
+    K8S_FEATURE_GATES="ReadWriteOncePod=true"
```
This will override other feature gates if the user has passed them.

humblec replied:
@Rakshith-R addressed. ptal. thanks

Rakshith-R commented on e2e/cephfs.go:
```diff
@@ -345,6 +345,61 @@ var _ = Describe("cephfs", func() {
 			})
 
+			By("verify RWOP volume support", func() {
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
+					Skip("RWOP PVC access mode is only supported from v1.22+")
+				}
```
```suggestion
				if k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
					err=ValidateCephFSRWOP()
					if err!=nil{
					...
					}
				}
```
@humblec , this will lead to entire ci being skipped 
refer https://github.com/ceph/ceph-csi/pull/2680 

Example https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/3467/pipeline/99

humblec replied:
ah. PR created before the referred one :) , will intend it properly.

Rakshith-R commented on e2e/rbd.go:
```diff
@@ -465,6 +467,98 @@ var _ = Describe("RBD", func() {
 				validateRBDImageCount(f, 0, defaultRBDPool)
 			})
+			By("create a Block mode RWOP PVC and bind it to more than one app", func() {
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
+					Skip("RWOP PVC access mode is only supported from v1.22+")
+				}
```
same here

Rakshith-R replied:
This is not changed yet.

humblec replied:
Done

Rakshith-R commented on README.md:
```diff
@@ -81,7 +81,9 @@ for its support details.
 |        | Dynamically provision, de-provision Block mode RWX volume | GA             | >= v1.0.0          | >= v1.0.0        | Nautilus (>=14.0.0)  | >= v1.14.0         |
+|        | Dynamically provision, de-provision Block mode RWOP volume| GA             | >= v3.5.0          | >= v1.5.0        | Nautilus (>=14.0.0)  | >= v1.22.0         |
 |        | Dynamically provision, de-provision File mode RWO volume  | GA             | >= v1.0.0          | >= v1.0.0        | Nautilus (>=14.0.0)  | >= v1.14.0         |
+|        | Dynamically provision, de-provision File mode RWOP volume | GA             | >= v3.5.0          | >= v1.5.0        | Nautilus (>=14.0.0)  | >= v1.22.0         |
```
Add The same line for cephfs too?

humblec replied:
Done.

Rakshith-R commented on e2e/rbd.go:
```diff
@@ -465,6 +467,98 @@ var _ = Describe("RBD", func() {
+				validateRBDImageCount(f, 0, defaultRBDPool)
+			})
+			By("create a RWOP PVC and bind it to more than one app", func() {
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
+					Skip("RWOP PVC access mode is only supported from v1.22+")
+				}
```
same here.

This probably can be made from modular by passing in pvc & app path as parameters to be re-used for rbd block and file mode e2e.

humblec replied:
sure.

Rakshith-R reviewed (dismissed):
@humblec  ci is being skipped for k8s <1.22
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/3467/pipeline/99
refer https://github.com/ceph/ceph-csi/pull/2680

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e/k8s-1.23

humblec commented:
/retest ci/centos/upgrade-tests-rbd

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -94,25 +94,52 @@ func (cs *ControllerServer) validateVolumeReq(ctx context.Context, req *csi.Crea
+// if the capability is of multi write the first return value will be set to true and if the request
+// is of type block, the second return value will be set to true. If it is of type RWO and file mode, the third return
+// will be set to true. Otherwise, false values are returned.
+func isMultiWriterBlockRWOFile(caps []*csi.VolumeCapability) (bool, bool, bool, bool) {
```
Move this function to csi-common utils.

nixpanic replied:
Four return variables is a LOT, please document them well.

nixpanic replied:
See #2717 for a similar change that helps with implementing CSI-Addons ReclaimSpace.

humblec replied:
rearranged this function to csi commons utils.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -94,25 +94,52 @@ func (cs *ControllerServer) validateVolumeReq(ctx context.Context, req *csi.Crea
+// is of type block, the second return value will be set to true. If it is of type RWO and file mode, the third return
+// will be set to true. Otherwise, false values are returned.
+func isMultiWriterBlockRWOFile(caps []*csi.VolumeCapability) (bool, bool, bool, bool) {
+	var multiWriter, Block, rwoFile, roOnly bool
```
why is `Block` starting with a capital `B`?

Maybe place all vars on their own line, so that they can be commented? The naming is not extremely telling what they mean.

humblec replied:
yeah,not  need to be in caps, changed to `b` and also splitted the declaration of the vars to seperate lines with more explanation on each of them.

nixpanic commented on examples/rbd/raw-block-pvc-rwop.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: raw-block-pvc
```
this name is already used, please name it something else

humblec replied:
done

nixpanic replied:
I do still see the `raw-block-pvc` name here, and `rbd-pvc` in the `pvc-rwop.yaml` is already used too. Please use a name for the PVCs that contain `rwop` too.

nixpanic commented on examples/cephfs/pvc-rwop.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: csi-cephfs-pvc
```
better use a different name here too

humblec replied:
Done

humblec replied:
@nixpanic one thing I noticed here was, if I change the name in the yaml, I have to introduce 3 other yamls ( POD yamls) for new these PVCs. if really required I can add that, otherwise we can stick to the default POD yamls with these name. is that fine ?

nixpanic replied:
yes, that is really required. Use a name for the PVC that describes the `rwop` part. It is very confusing when users create the PVCs manually from these examples. Examples need to have descriptive names, otherwise we would use `pvc` and `pod` and `storageclass` for everything else too.

nixpanic reviewed (dismissed):
just a few nits, looks good to me otherwise

humblec commented:
> just a few nits, looks good to me otherwise

Thanks @nixpanic , let me address the comments. :+1:

humblec commented:
@nixpanic comments are addressed. ptal.. thanks :+1:  Cc @Rakshith-R

humblec commented:
@Mergifyio rebase

humblec commented:
#2728

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
@nixpanic @Rakshith-R please review :)

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
Really appreciate a review on this as this has been marked for release-3.5 . Thanks.

Rakshith-R commented:
@humblec , please consider increasing the timeout, refer #2739 for more infor

humblec commented:
This test add extra ~15 minutes to the test run, because it is on the failure scenario of second pod and thus 3 tests, will decrease the timeout of error and check .

humblec commented:
> This test add extra ~15 minutes to the test run, because it is on the failure scenario of second pod and thus 3 tests, will decrease the timeout of error and check .

This has been addressed now without increasing the timeout. Tests are passing too. ptal with priority please. thanks..

Madhu-1 commented on README.md:
```diff
@@ -96,6 +98,7 @@ for its support details.
 | CephFS | Dynamically provision, de-provision File mode RWO volume  | Beta           | >= v1.1.0          | >= v1.0.0        | Nautilus (>=14.2.2)  | >= v1.14.0         |
 |        | Dynamically provision, de-provision File mode RWX volume  | Beta           | >= v1.1.0          | >= v1.0.0        | Nautilus (>=v14.2.2) | >= v1.14.0         |
 |        | Dynamically provision, de-provision File mode ROX volume  | Alpha          | >= v3.0.0          | >= v1.0.0        | Nautilus (>=v14.2.2) | >= v1.14.0         |
+|        | Dynamically provision, de-provision File mode RWOP volume | GA             | >= v3.5.0          | >= v1.5.0        | Nautilus (>=14.0.0)  | >= v1.22.0         |
```
why mark it as GA in the first PR?

humblec replied:
marked it as alpha for now. thanks.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -35,7 +35,6 @@ var (
 
 func deployCephfsPlugin() {
 	// delete objects deployed by rook
-
```
remove this change

humblec replied:
done.

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -344,6 +343,60 @@ var _ = Describe("cephfs", func() {
+						e2elog.Failf("failed to load application: %v", err)
+					}
+					app.Namespace = f.UniqueName
+					aname := app.Name
```
`aname` can have a better naming

humblec replied:
done

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -344,6 +343,60 @@ var _ = Describe("cephfs", func() {
+					}
+					validateSubvolumeCount(f, 1, fileSystemName, subvolumegroup)
+
+					// create one more  app with same PVC
```
`// create one more  app with same PVC` to `// create one more app with same PVC`

humblec replied:
done

Madhu-1 commented on e2e/cephfs.go:
```diff
@@ -344,6 +343,60 @@ var _ = Describe("cephfs", func() {
+					app.Name = name
+					err = createAppErr(f.ClientSet, app, deployTimeout, ErrRWOPConflict)
+					if err != nil {
+						e2elog.Failf("application should not go to running state due to RWOP access mode")
```
log error message.

humblec replied:
done

Madhu-1 commented on e2e/errors.go:
```diff
@@ -23,6 +23,8 @@ import (
 	utilnet "k8s.io/apimachinery/pkg/util/net"
 )
 
+const ErrRWOPConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode."
```
why its exported?

humblec replied:
unexported

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -487,6 +489,96 @@ var _ = Describe("RBD", func() {
 			})
+			By("create a Block mode RWOP PVC and bind it to more than one app", func() {
+				if k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
+					pvc, err := loadPVC(rawPVCRWOPPath)
+					if err != nil {
+						e2elog.Failf("failed to load PVC: %v", err)
+					}
+					pvc.Namespace = f.UniqueName
+
+					app, err := loadApp(rawAppPath)
+					if err != nil {
+						e2elog.Failf("failed to load application: %v", err)
+					}
+					app.Namespace = f.UniqueName
+					aname := app.Name
+					err = createPVCAndvalidatePV(f.ClientSet, pvc, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to create PVC: %v", err)
+					}
+					// validate created backend rbd images
+					validateRBDImageCount(f, 1, defaultRBDPool)
+
+					err = createApp(f.ClientSet, app, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to create application: %v", err)
+					}
+					// create one more  app with same PVC
+					name := fmt.Sprintf("%s%d", f.UniqueName, deployTimeout)
+					app.Name = name
+					err = createAppErr(f.ClientSet, app, deployTimeout, ErrRWOPConflict)
+					if err != nil {
+						e2elog.Failf("application should not go to running state due to RWOP access mode")
+					}
+					err = deletePod(name, app.Namespace, f.ClientSet, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to delete application: %v", err)
```
the logic to validate RWOP remains the same for both rbd and cephfs can we have a common function for both to avoid code duplication?

humblec replied:
done.

Madhu-1 commented on internal/csi-common/utils.go:
```diff
@@ -306,3 +306,46 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
+			// however more granular level check could help us in future, so keeping it here as an additional measure.
+			if cap.GetMount() != nil {
+				switch cap.AccessMode.Mode { //nolint:exhaustive // only check what we want
+				case csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER:
```
instead of having more values in single case, have multiple cases statements with fallthrough?

humblec replied:
kept this model as it is a common standard as well. thanks..

Madhu-1 commented:
```bash
[root@dhcp53-170 rbd]# kubectl get po/csi-rbdplugin-provisioner-64886b44d5-wb5zj -nrook-ceph -oyaml |grep -i cephcsi
    image: quay.io/cephcsi/cephcsi:v3.4.0
    image: quay.io/cephcsi/cephcsi:v3.4.0
    image: quay.io/cephcsi/cephcsi:v3.4.0
    image: quay.io/cephcsi/cephcsi:v3.4.0
    imageID: docker-pullable://quay.io/cephcsi/cephcsi@sha256:1a6b395ffed6e51b7b73a87694690283d91b782b56cfaafee447f263c68a55d9
    image: quay.io/cephcsi/cephcsi:v3.4.0
    imageID: docker-pullable://quay.io/cephcsi/cephcsi@sha256:1a6b395ffed6e51b7b73a87694690283d91b782b56cfaafee447f263c68a55d9
    image: quay.io/cephcsi/cephcsi:v3.4.0
    imageID: docker-pullable://quay.io/cephcsi/cephcsi@sha256:1a6b395ffed6e51b7b73a87694690283d91b782b56cfaafee447f263c68a55d9

[root@dhcp53-170 rbd]# kubectl get pvc
NAME                STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS      AGE
rbd-pvc-demo-rwop   Bound    pvc-096fcd6c-ff1f-463a-ae05-3a8f57a317e9   1Gi        RWOP           rook-ceph-block   64s
rbd-pvc-test1       Bound    pvc-7783127f-a758-4ee9-8eac-4b813f61be69   8Gi        RWO            rook-ceph-block   39h
[root@dhcp53-170 rbd]# kubectl get po
NAME               READY   STATUS    RESTARTS   AGE
csi-rbd-demo-pod   1/1     Running   0          52s


[root@dhcp53-170 rbd]# kubectl create -f pod.yaml 
pod/csi-rbd-demo-pod-0 created
[root@dhcp53-170 rbd]# kubectl describe pod/csi-rbd-demo-pod-0
Name:         csi-rbd-demo-pod-0
Namespace:    default
Priority:     0
Node:         <none>
Labels:       <none>
Annotations:  <none>
Status:       Pending
IP:           
IPs:          <none>
Containers:
  web-server:
    Image:        nginx
    Port:         <none>
    Host Port:    <none>
    Environment:  <none>
    Mounts:
      /var/lib/www/html from mypvc (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-b795q (ro)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  mypvc:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  rbd-pvc-demo-rwop
    ReadOnly:   false
  kube-api-access-b795q:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  24s   default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

@humblec without this PR am getting the expected behaviour what exactly this PR is allowing/blocking?

humblec committed:
doc: add RWOP support matrix to README.md

This commit add RWOP support matrix to README

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

humblec commented:
@humblec without this PR am getting the expected behaviour what exactly this PR is allowing/blocking?

@Madhu-1 This PR is about making the CSI driver aware and advertise the caps introduced in  CSI spec v1.5 as mentioned in the PR description [1]. This PR also added the RWOP validation E2E tests here in this repo. With new caps check or advertisement we havent broken any existing functionality either.  As we have updated the CSI spec to be on version 1.5 this is introduced.   This also helps us or set the platform to do some restrictions on mount operation (ex: iscsi) if needed later.  

[1] # https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/#as-a-storage-vendor-how-do-i-add-support-for-this-access-mode-to-my-csi-driver

Please let me know if you have any further queries on this. 

Also, the changes requested has been addressed, so appreciate if you can revisit . 

@ceph/ceph-csi-contributors  Considering this has been in review for long time and all the comments are addressed and marked for 3.5 release, can you review this PR with priority?

Madhu-1 commented:
> @humblec without this PR am getting the expected behaviour what exactly this PR is allowing/blocking?
> 
> @Madhu-1 This PR is about making the CSI driver aware and advertise the caps introduced in CSI spec v1.5 as mentioned in the PR description [1]. This PR also added the RWOP validation E2E tests here in this repo. With new caps check or advertisement we havent broken any existing functionality either. As we have updated the CSI spec to be on version 1.5 this is introduced. This also helps us or set the platform to do some restrictions on mount operation (ex: iscsi) if needed later.
> 
> [1] # https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/#as-a-storage-vendor-how-do-i-add-support-for-this-access-mode-to-my-csi-driver
> 

I don't see any validation as mentioned at https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/#implement-nodepublishvolume-behavior to make sure cephcsi is not blocking publish to more target paths.

> Please let me know if you have any further queries on this.
> 

IMO we don't add anything without a valid reason, adding more things thinking something might be needed for later is not a valid justification. it can be added when it's required. 


> Also, the changes requested has been addressed, so appreciate if you can revisit .
> 

Removing changes requests if someone thinks it's useful and approves to merge it.

> @ceph/ceph-csi-contributors Considering this has been in review for long time and all the comments are addressed and marked for 3.5 release, can you review this PR with priority?

humblec commented:
> > @humblec without this PR am getting the expected behaviour what exactly this PR is allowing/blocking?
> > @Madhu-1 This PR is about making the CSI driver aware and advertise the caps introduced in CSI spec v1.5 as mentioned in the PR description [1]. This PR also added the RWOP validation E2E tests here in this repo. With new caps check or advertisement we havent broken any existing functionality either. As we have updated the CSI spec to be on version 1.5 this is introduced. This also helps us or set the platform to do some restrictions on mount operation (ex: iscsi) if needed later.
> > [1] # https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/#as-a-storage-vendor-how-do-i-add-support-for-this-access-mode-to-my-csi-driver
> 
> I don't see any validation as mentioned at https://v1-22.docs.kubernetes.io/blog/2021/09/13/read-write-once-pod-access-mode-alpha/#implement-nodepublishvolume-behavior to make sure cephcsi is not blocking publish to more target paths.
> 
As discussed above, its not the goal of this PR.

> > Please let me know if you have any further queries on this.
> 
> IMO we don't add anything without a valid reason, adding more things thinking something might be needed for later is not a valid justification. it can be added when it's required.

Again, this PR is not just about the validation/blocking or accepting the request. Half of the commits are adding RWOP support validation in e2e. 
A couple of them advertising the caps....etc. 
Also please note making CSI driver to be compatible with the spec is also a practice, for ex:  https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/797

> 
> > Also, the changes requested has been addressed, so appreciate if you can revisit .
> 
> Removing changes requests if someone thinks it's useful and approves to merge it.
> 

Thanks!

nixpanic commented on internal/csi-common/utils.go:
```diff
@@ -320,3 +320,46 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
+// will be set to true. If the access mode is reader only the last return value has been set to true.
+// Otherwise, false values are returned for all the return arguments.
+func IsMultiWriterBlockRWOFile(caps []*csi.VolumeCapability) (bool, bool, bool, bool) {
+	var multiWriter bool // this flag has been set and returned after validating multi writer caps regardless of
```
due to all the comments, the formatting is getting rather difficult to follow. It might be better to stick to the commenting above variables, similar to what is done for functions and members of a struct.

```go
var (
    // multiWriter has been set and returned after validating multi writer caps regardless of
    // single or multi node access mode. The caps check is agnostic to whether it is a filesystem
    // or block mode volume
    multiWriter bool
                                                      // <--- NOTE THE EMPTY LINE FOR READABILTY!!!
    // block has been set and returned if the passed in capability is of block volume mode
    block bool
)
```

humblec replied:
Done. actually it was not needed as part of function split

nixpanic commented on internal/csi-common/utils.go:
```diff
@@ -320,3 +320,46 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
 	return isBlock, isMultiNode
 }
+
+// IsMultiWriterBlockRWOFile validates the volume capability slice against the access modes and access type.
```
This function is horrible, it returns **FOUR** booleans. It would be much easier to read when the function is split in multiple functions, each returning their own `isBlock()`, `isMultiWriter()` and so on. There are very few capabilities in the slice that is passed to this function, there is no need to optimize for looping fewer times. Just have the caller call `isMultiWriter()` where it needs. It will improve the maintainability _so much_.

humblec replied:
Done. splitted into different functions.

nixpanic commented on internal/csi-common/utils_test.go:
```diff
@@ -177,3 +177,219 @@ func TestIsBlockMultiNode(t *testing.T) {
 	}
 }
+
+func TestIsMultiWriterBlockRWOFile(t *testing.T) {
```
The size of this unit-test should have suggested that `IsMultiWriterBlockRWOFile()` is not a very practical function to use. Make the code simpler by splitting `IsMultiWriterBlockRWOFile()` in separate functions please.

humblec replied:
Done. segregated into different functions.

nixpanic commented on e2e/pod.go:
```diff
@@ -414,3 +414,30 @@ func recreateCSIPods(f *framework.Framework, podLabels, daemonsetName, deploymen
+	err = createAppErr(f.ClientSet, app, deployTimeout, errRWOPConflict)
+	if err != nil {
+		return fmt.Errorf("application should not go to running state due to RWOP access mode: %w", err)
+	}
```
Add an empty line after the closing `}` from if-statements please. This function reads very difficult.

humblec replied:
Done.

humblec replied:
Done.

nixpanic reviewed (changes requested):
just a few small improvements to improve code maintenance

humblec committed:
e2e: add example yamls for RWOP POD and PVC

This commit add example yamls for RBD and CephFS Pod and PVCs

RBD:
 Raw Block Volume and File Mode PVCs with RWOP accessmode
 Raw Block Volume POD and FileMode POD yamls referring RWOP PVC

CephFS:
 RWOP PVC and POD yaml referring RWOP PVC

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: add rwop validation helper function in pod.go

This commit adds the rwop validation helper for rbd and cephfs tests.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: add rwop validation tests for rbd filemode and block mode

This commit adds the validation tests for rbd block and filemode
PVCs.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec committed:
e2e: add rwop validation for cephfs volumes

This commit adds the rwop validation tests for cephfs.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

nixpanic commented:
Linting the `.yaml` files seem to fail. You still need to update the names of the PVCs too.

humblec commented:
> Linting the `.yaml` files seem to fail. You still need to update the names of the PVCs too.

@nixpanic added the e2e part too.. all comments should be addressed by now. Thanks :+1:

nixpanic commented on internal/csi-common/utils.go:
```diff
@@ -320,3 +320,74 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
+				case csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_MULTI_WRITER,
+					csi.VolumeCapability_AccessMode_SINGLE_NODE_SINGLE_WRITER:
+					rwoFile = true
```
you can return `true` here, no need to continue with the for-loop

humblec replied:
Done..thanks!

nixpanic commented on internal/csi-common/utils.go:
```diff
@@ -320,3 +320,74 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
+		}
+	}
+
+	return rwoFile
```
if you return `true` above, just return `false` here. No need for the `rwoFile` variable.

humblec replied:
Done

nixpanic commented on internal/csi-common/utils.go:
```diff
@@ -320,3 +320,74 @@ func IsBlockMultiNode(caps []*csi.VolumeCapability) (bool, bool) {
+			switch cap.AccessMode.Mode { //nolint:exhaustive // only check what we want
+			case csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY,
+				csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY:
+				roOnly = true
```
similar here, just return `true` here, and drop the `roOnly` variable

humblec replied:
Done

nixpanic reviewed (changes requested):
Thanks, looks much easier to maintain now!

Just two small cleanups, and you'll get the +1 from me :+1:

humblec commented:
> Thanks, looks much easier to maintain now!

 Cool :+1: 
> Just two small cleanups, and you'll get the +1 from me +1

Sure, addressed those too! Thanks @nixpanic

humblec commented:
RWOP tests are passing as expected and also quick: (~1m)
```
[1mSTEP[0m: create a Block mode RWOP PVC and bind it to more than one app
Jan  6 18:07:17.522: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{raw-block-rwop-pvc  rbd-1318    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[
...........
Jan  6 18:07:42.697: INFO: Expected Error "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode." found successfully
Jan  6 18:07:42.712: INFO: Waiting for pod rbd-131810 to be deleted
Jan  6 18:07:42.727: INFO: Waiting for pod csi-rbd-demo-rwop-pod to be deleted
Jan  6 18:08:14.739: INFO: Deleting PersistentVolumeClaim raw-block-rwop-pvc on namespace rbd-1318
Jan  6 18:08:14.752: INFO: waiting for PVC raw-block-rwop-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOncePod],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,} to be deleted (0 seconds elapsed)
Jan  6 18:08:16.756: INFO: waiting for PVC raw-block-rwop-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOncePod],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,} to be deleted (2 seconds elapsed)
Jan  6 18:08:16.769: INFO: ExecWithOptions {Command:[/bin/sh -c rbd ls --format=json --pool=replicapool] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-hnvgz ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Jan  6 18:08:16.769: INFO: >>> kubeConfig: /root/.kube/config
Jan  6 18:08:16.770: INFO: ExecWithOptions: Clientset creation
Jan  6 18:08:16.770: INFO: ExecWithOptions: execute(POST https://192.168.39.79:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7467d8bf8-hnvgz/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+ls+--format%3Djson+--pool%3Dreplicapool&container=rook-ceph-tools&container=rook-ceph-tools&stderr=true&stdout=true %!s(MISSING))
[1mSTEP[0m: create a RWOP PVC and bind it to more than one app
.....
Jan  6 18:08:40.963: INFO: Expected Error "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode." found successfully
Jan  6 18:08:40.974: INFO: Waiting for pod rbd-131810 to be deleted
Jan  6 18:08:40.984: INFO: Waiting for pod csi-rbd-demo-fs-rwop-pod to be deleted
Jan  6 18:08:42.992: INFO: Deleting PersistentVolumeClaim rbd-rwop-pvc on namespace rbd-1318
Jan  6 18:08:43.002: INFO: waiting for PVC rbd-rwop-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOncePod],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,} to be deleted (0 seconds elapsed)
Jan  6 18:08:45.006: INFO: waiting for PVC rbd-rwop-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOncePod],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,} to be deleted (2 seconds elapsed)
..
Jan  6 18:08:45.017: INFO: >>> kubeConfig: /root/.kube/config
Jan  6 18:08:45.018: INFO: ExecWithOptions: Clientset creation
Jan  6 18:08:45.018: INFO: ExecWithOptions: execute(POST https://192.168.39.79:8443/api/v1/namespaces/rook-ceph/pods/rook-ceph-tools-7467d8bf8-hnvgz/exec?command=%2Fbin%2Fsh&command=-c&command=rbd+ls+--format%3Djson+--pool%3Dreplicapool&container=rook-ceph-tools&container=rook-ceph-tools&stderr=true&stdout=true %!s(MISSING))
[1mSTEP[0m: create an erasure coded PVC and bind it to an app
````

NBD tests looks to be failing as we have seen earlier frequently.

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented on internal/csi-common/utils_test.go:
```diff
@@ -462,7 +462,6 @@ func TestIsBlockMultiWriter(t *testing.T) {
 			if block != newtt.block {
 				t.Errorf("IsBlockMultiWriter block = %v, want %v", block, newtt.block)
 			}
-
```
this deletion of the line should probably be moved to a different commit.

nixpanic commented on scripts/minikube.sh:
```diff
@@ -203,6 +212,14 @@ up)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 22 ];
+    then
+    # if kubernetes version is greater than 1.22 enable RWOP feature gate
+    K8S_FEATURE_GATES="${K8S_FEATURE_GATES},ReadWriteOncePod=true"
```
indention is off?

humblec replied:
done.

nixpanic replied:
no, I don't thinks so?

humblec replied:
@nixpanic looks like I missed to commit, done now.. Thanks!

nixpanic commented on e2e/errors.go:
```diff
@@ -23,6 +23,8 @@ import (
 	utilnet "k8s.io/apimachinery/pkg/util/net"
 )
 
+const errRWOPConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode."
```
This is a string, not an Error. Why is it in this file if it is never used here?

humblec replied:
sure, moved to proper place.

nixpanic commented on e2e/rbd.go:
```diff
@@ -505,6 +509,72 @@ var _ = Describe("RBD", func() {
 				// validate created backend rbd images
 				validateRBDImageCount(f, 0, defaultRBDPool)
 			})
+			By("create a Block mode RWOP PVC and bind it to more than one app", func() {
```
missing empty line before `By`?

humblec replied:
Done

nixpanic reviewed (commented):
Few nits on the coding standard, it would be good to address those. No functional or other issues that I see.

humblec committed:
deploy: adjust minikube deployment for RWOP support

The ReadWriteOncePod feature gate need to be enabled only when we
are operating on kube 1.22 or above cluster. This commit adds the
logic to parse the kubernetes cluster at time of minikube deployment
and if it is above v1.22, enable the RWOP feature gate

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
> Few nits on the coding standard, it would be good to address those. No functional or other issues that I see.

@nixpanic addressed final nits.. thanks.. ptal.. :+1:

nixpanic commented:
> > Few nits on the coding standard, it would be good to address those. No functional or other issues that I see.
> 
> @nixpanic addressed final nits.. thanks.. ptal.. +1

not sure, but I do not see how you addressed https://github.com/ceph/ceph-csi/pull/2359#discussion_r780184642 ?

yati1998 reviewed (approved):
LGTM

humblec commented:
/retest ci/centos/k8s-e2e-external-storage/1.23

humblec commented:
@Mergifyio refresh

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio rebase

humblec commented:
@Mergifyio refresh
